### PR TITLE
[pytorch][nnc] external function call to xnnpack ops

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59525 [pytorch][nnc] external function call to xnnpack ops**

This PR added NNC external function call binding for two XNNPack ops:
- prepacked::linear_clamp_run
- prepacked::conv2d_clamp_run

Both ops take two arguments: a regular input tensor and a prepacked context
object that contains other parameters like weights/bias/etc. The prepacked
context object's type is a custom class.

NNC doesn't generate assembly code that reads the content of the prepacked
object directly. It simply passes it into the XNNPack ops wrapper, so both
NNC and the generated assembly code don't need to know the custom class type.

At compilation time, we use a size-1 dummy tensor as the placeholder for the
prepacked XNNPack context object.

At runtime, we pass in the raw pointer of the XNNPack context object as if it
were a regular tensor storage data pointer.

Inside the external function call wrapper, we reinterpret_cast the raw pointer
back to the custom class type before dispatching to the XNNPack ops.

Differential Revision: [D28924934](https://our.internmc.facebook.com/intern/diff/D28924934/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28924934/)!

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/external_functions.cpp [hunk 1]
```diff
@@ -181,6 +183,66 @@ void nnc_aten_triangular_solve(
...
+      c10::intrusive_ptr<LinearOpContext>::unsafe_reclaim_from_nonowning(
+          reinterpret_cast<LinearOpContext*>(buf_data[2]));
+
+  static auto op =
+      c10::Dispatcher::singleton()
+          .findSchemaOrThrow("prepacked::linear_clamp_run", "")
+          .typed<at::Tensor(
+              const at::Tensor&, const c10::intrusive_ptr<LinearOpContext>&)>();
+  at::Tensor output = op.call(x, prepacked);
+  memcpy(
```
Comment (kimishpatel) [inline torch/csrc/jit/tensorexpr/external_functions.cpp:212]: Intersting, that memcpy is not costing you perf. But I am wondering if you can just do `buf_data[0] = output.data_ptr()`. Maybe not exactly in that form and something rather does exchange data pointers and frees appropriately.

File: torch/csrc/jit/tensorexpr/external_functions.cpp [hunk 2]
```diff
@@ -181,6 +183,66 @@ void nnc_aten_triangular_solve(
...
+  std::vector<at::Tensor> tensors =
+      constructTensors(bufs_num - 1, buf_data, buf_ranks, buf_dims, buf_dtypes);
+
+  const at::Tensor& x = tensors[1];
+  const auto prepacked =
+      c10::intrusive_ptr<LinearOpContext>::unsafe_reclaim_from_nonowning(
+          reinterpret_cast<LinearOpContext*>(buf_data[2]));
+
+  static auto op =
+      c10::Dispatcher::singleton()
```
Comment (kimishpatel) [inline torch/csrc/jit/tensorexpr/external_functions.cpp:207]: Why do you need to go through dispatcher. I suppose in a generic case you do need to know which op to invoke?

File: test/cpp/tensorexpr/test_external_calls.cpp [hunk 3]
```diff
@@ -249,6 +251,162 @@ TEST(ExternalCall, Addmm_float) {
...
+
+#ifdef TORCH_ENABLE_LLVM
+  LLVMCodeGen llvm_codegen(l.root_stmt(), {Input, DummyPrepacked, Result});
+
+  llvm_codegen.call({input_buf, prepacked.get(), result_buf});
+  nnc_result = at::from_blob(result_buf.data(), {100, 300}, options);
+  ASSERT_TRUE(at::allclose(nnc_result, ref));
+#endif
+
+  SimpleIREvaluator ir_eval(l.root_stmt(), {Input, DummyPrepacked, Result});
```
Comment (kimishpatel) [inline test/cpp/tensorexpr/test_external_calls.cpp:314]: What is the difference between codegen and ir_eval?

**Review by kimishpatel — COMMENTED** [review]
Left some comments.

File: torch/csrc/jit/tensorexpr/external_functions.cpp (hunk 2 above)
Comment (ljk53) [reply torch/csrc/jit/tensorexpr/external_functions.cpp]: > Why do you need to go through dispatcher. I suppose in a generic case you do need to know which op to invoke?

@kimishpatel my original thought was to keep using dispatcher before we figure out how to call third-party/xnnpack library directly without going through the aten/native/xnnpack wrappers. I didn't want to have much dependency on aten/native/... stuff directly as they are generally considered "hidden" implementation details.

For clients like static runtime where we don't want to incur dispatcher overhead, I think we generate separate public header functions for overhead-avoidance purpose. Unfortunately we don't have equivalent thing for manually registered custom ops - I don't want to set a "bad" example here :)

After inspecting the code a bit more, I changed the plan a little bit. Seems these XNNPack ops are very special - they simply call the "context->run(input)" without doing any other pre/post processing. The existing pattern is to wrap all the pre/post processing logic inside the LinearOpContext/Conv2dOpContext classes, which we have to expose anyway. So here I changed it to call "context->run()" directly as well. The risk is that if we ever change the aten side linear_clamp_run() / conv2d_clamp_run() implementation to do more work than "context->run()", then we need to update these external function wrappers here as well. In other words, "LinearOpContext" and "Conv2dOpContext" methods become "public" APIs that the op developers need to be aware of.

File: torch/csrc/jit/tensorexpr/external_functions.cpp (hunk 1 above)
Comment (ljk53) [reply torch/csrc/jit/tensorexpr/external_functions.cpp:213]: Maybe we can add a "context->run_out(input, buf_data[0])" method to write the result into the output buffer directly (we also need to update the buffer size calculation logic to consider the extra memory padding required by the XNNPack).

> if you can just do `buf_data[0] = output.data_ptr()`. Maybe not exactly in that form and something rather does exchange data pointers and frees appropriately.

We cannot simply update the pointer - as you pointed out, the "frees appropriately" part is not trivial. The "output" returned from the aten is a c++ object, which needs to be destructed properly. Right now we manage the life-cycle of all intermediate buffers outside of the assembly function - with this approach we can avoid handling object allocation/free in the assembly code.

Currently all these intermediate buffers are not constructed as "real" aten tensors - the caller simply does "malloc(size)" and "free(buf)". We don't need to ever construct "real" aten tensors for ops that NNC can generate kernels without calling aten kernels (which should become the majority cases). If we want to construct/destruct these intermediate buffers as real tensors, then we need to change the protocol to track more metadata about the tensors (dtype, shape, etc). It's possible but will take some effort.

Maybe eventually we should handle these inside the assembly code (especially when we add support for dynamic shaped inputs), and we will need to have some caching allocator support to avoid memory management cost.

File: test/cpp/tensorexpr/test_external_calls.cpp [hunk 4]
```diff
@@ -249,6 +251,162 @@ TEST(ExternalCall, Addmm_float) {
...
+
+  // Calculate reference result using at::linear.
+  auto options = at::TensorOptions()
+                     .dtype(at::kFloat)
+                     .layout(at::kStrided)
+                     .device(at::kCPU)
+                     .requires_grad(false);
+  at::Tensor input = at::ones({100, 200}, options) * 5.f;
+  at::Tensor weight = at::ones({300, 200}, options) * 6.f;
+  at::Tensor bias = at::ones({300}, options) * 11.f;
```
Comment (raziel) [inline test/cpp/tensorexpr/test_external_calls.cpp:275]: these are not needed

also, does the input need to have the right shapes if it's meant as a placeholder?

File: test/cpp/tensorexpr/test_external_calls.cpp [hunk 5]
```diff
@@ -249,6 +251,162 @@ TEST(ExternalCall, Addmm_float) {
...
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST(ExternalCall, Prepacked_Conv2d_float) {
+  using namespace at::native::xnnpack;
+
+  KernelScope kernel_scope;
+
+  Placeholder Input("Input", kFloat, {1, 3, 224, 224});
+  Placeholder Weight("Weight", kFloat, {16, 3, 3, 3});
+  Placeholder Bias("Bias", kFloat, {16});
```
Comment (raziel) [inline test/cpp/tensorexpr/test_external_calls.cpp:329]: Same comment here as above.

File: torch/csrc/jit/tensorexpr/external_functions.cpp (hunk 1 above)
Comment (raziel) [reply torch/csrc/jit/tensorexpr/external_functions.cpp:213]: "Maybe we can add a "context->run_out(input, buf_data[0])" method to write the result into the output buffer directly (we also need to update the buffer size calculation logic to consider the extra memory padding required by the XNNPack)."

Do you think we'll make all these calls always of "out variant" type? Then make the allocation part of a memory planning stage? Probably then also be more than simple functions (e.g. in the future may need their own shape computation function)?

File: test/cpp/tensorexpr/test_external_calls.cpp (hunk 4 above)
Comment (ljk53) [reply test/cpp/tensorexpr/test_external_calls.cpp]: deleted - right, I don't think the input shape has to be real for this case.

File: test/cpp/tensorexpr/test_external_calls.cpp (hunk 5 above)
Comment (ljk53) [reply test/cpp/tensorexpr/test_external_calls.cpp]: deleted

File: torch/csrc/jit/tensorexpr/external_functions.cpp (hunk 1 above)
Comment (ljk53) [reply torch/csrc/jit/tensorexpr/external_functions.cpp:213]: > Do you think we'll make all these calls always of "out variant" type? Then make the allocation part of a memory planning stage? Probably then also be more than simple functions (e.g. in the future may need their own shape computation function)?

Yes I think so. The shape calculation will be more than simple functions when we support dynamic shaped inputs. I don't know whether we want to manage these allocations outside of the assembly code or inside the assembly code.

To do it outside of the assembly code would require some DSL (or maybe with torchscript wrapper?) to specify the shape calculation function for each intermediate buffer.

To do it inside the assembly code we can either directly generate the shape calculation function, or insert a call to the meta-function of structured kernels: https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md. And I imagine we will need to integrate with the caching allocator to avoid repetitive allocations for continuous inferencing.

Comment (ljk53) [reply torch/csrc/jit/tensorexpr/external_functions.cpp:213]: To be clear, doing it inside the assembly code (i.e. allocating buffer inline + calling out-variant) would be similar as directly calling the functional-variant. It's just that the current combination (allocating buffers outside + calling functional-variant + copy data) doesn't make a lot of sense :)

Comment (kimishpatel) [reply torch/csrc/jit/tensorexpr/external_functions.cpp:213]: I would imagine that you still want to do output allocations outside of the kernel assembly code, mainly so as to allow memory planning be separate from kernels. In this particular case maybe an out variant of the op can provided and used to avoid copying data.

Note that the advantage of using memory allocations outside of the assembly is: 1) we can do liveness analysis and reuse memory as needed. 2) A common memory planner shared between compiler runtime and pt core runtime might be usable.
Main trouble I see with this approach is when allocations are input dependent, such as in maskrcnn where rois are variable.
